validator/client/beacon-api/index.go

```diff
@@ -0,0 +1,82 @@
+//go:build use_beacon_api
+// +build use_beacon_api
+
+package beacon_api
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	neturl "net/url"
+	"strconv"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/api/gateway/apimiddleware"
+	rpcapi "github.com/prysmaticlabs/prysm/v3/beacon-chain/rpc/apimiddleware"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+)
+
+func (c beaconApiValidatorClient) validatorIndex(in *ethpb.ValidatorIndexRequest) (*ethpb.ValidatorIndexResponse, error) {
+	stringPubKey := hexutil.Encode(in.PublicKey)
+
+	stateValidator, err := c.getStateValidator(stringPubKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get validator state")
+	}
+
+	if len(stateValidator.Data) == 0 {
+		return nil, errors.Errorf("could not find validator index for public key `%s`", stringPubKey)
+	}
+
+	stringValidatorIndex := stateValidator.Data[0].Index
+
+	index, err := strconv.ParseUint(stringValidatorIndex, 10, 64)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot convert `%s` into uint64", stringValidatorIndex)
+	}
+
+	return &ethpb.ValidatorIndexResponse{Index: types.ValidatorIndex(index)}, nil
+}
+
+func (c beaconApiValidatorClient) getStateValidator(pubkey string) (rpcapi.StateValidatorsResponseJson, error) {
```

Comment:
I think this method should have its own method as well since it could be reused by other endpoints.

Reply:
Yes, it's the point of my "Comment for reviewers" section in PR description.

> Comments for reviewers:
Some code (like getStateValidator method) has a lot if common with the not yet merged https://github.com/prysmaticlabs/prysm/pull/11671 pull request.
Once this or https://github.com/prysmaticlabs/prysm/pull/11671 is merged, I will modify the not merged one to create generic methods where it is needed.

Reply:
I finally implemented a generic `getStateValidators` function in [this commit](https://github.com/prysmaticlabs/prysm/pull/11712/commits/4adf2b343d108df8c9ee9c2d405733c35d8d4508).